sljit_src/sljitProtExecAllocator.c

```diff
@@ -172,11 +182,11 @@ static SLJIT_INLINE int create_tempfile(void)
 #endif
 
 	if (tmp_name_len + 7 >= sizeof(tmp_name))
-	{
 		return -1;
-	}
 
-	strcpy(tmp_name + tmp_name_len, "/XXXXXX");
+	if (!(tmp_name_len == 1 && *tmp_name == '/'))
+		tmp_name[tmp_name_len++] = '/';
+	strcpy(tmp_name + tmp_name_len, "XXXXXX");
```

Comment:
I think `//XXXXXX` is valid, so we might not need to change this.

Reply:
still the new code avoids the unnecessary double "//" which might be problematic if this code ever gets ported to Novell or any other system where that makes a difference:

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_13

Reply:
this code is to revert the trimming of "/" that was done from the directory to use in lines 159-160 and that was technically not needed, since a directory with trailing "/" seems to work just fine for that open call (tested in Debian 10 with ext4 and tmpfs) as shown by:

 ```
#define _GNU_SOURCE
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>

int main(int argc, char *argv[])
{
  int fd = open(argv[1], O_TMPFILE|O_EXCL|O_RDWR|O_NOATIME|O_CLOEXEC, 0);
  if (fd < 1) {
    perror("temp file creation failed");
    exit(1);
  }

  close(fd);
  return 0;
}
```

note also that in modern Linux, that open call should succeed MOST (the only exception I can't think of are misconfigured docker containers) of the time since support for O_TMPFILE is wide spread and is definitely preferred to falling back to mkostemp, which has still a race condition in threaded mode.

either way, I think that trimming the trailing "/" and using only one "/" are cleaner so would rather have the code as proposed

Reply:
Maybe I was not clear.

The following code is correct:
```
if (tmp_name_len > 0 && tmp_name[tmp_name_len - 1] == '/')
		tmp_name[--tmp_name_len] = '\0';
strcpy(tmp_name + tmp_name_len, "/XXXXXX");
```

You added the extra stuff, because there is an `#ifdef` part which needs it. However this code might be not exist so we do extra work for nothing.

If we use the code inside the ifdef:
`fd = open(tmp_name_len > 0 ? tmp_name : "/", ...`

We get the best from both worlds.

Plus `if (tmp_name_len + 7 >= sizeof(tmp_name))` does not need to be updated as well.

Reply:
the first code has a bug, but as you correctly said it is mostly relevant outside the ifdef for O_TMPFILE, since it will break that call by passing "" as the directory.  you are correct that your proposed change avoids that but IMHO just works around the bug by making an even more unexpected (and undocumented) implementation.

the whole point of "refactoring" is to change the code, which is why I have to admit I am not sure if we should keep known bugs just to reduce the number of changed lines as an objective.

notice also that TMPDIR="/" is an edge case, and as I mentioned earlier the whole trimming of trailing "/" wasn't even needed (at least on my test system) but just because it works shouldn't be IMHO an incentive to not do the right thing.

finally; all this code could go away soon, because Linux could start using memfd_create (except for really old versions, and just because their libc didn't have the supported call, considering kernel support is ancient), and NetBSD >= 8 is already in a better/cleaner code path.

Reply:
You have correctly identified the bug, nice job! But there is a simpler (and I think nicer) solution to fix it.